File: 2024/14_Restroom_Redoubt/Day14.kt
// Day 14: Restroom Redoubt
// https://adventofcode.com/2024/day/14

import java.io.File
import kotlin.math.max
import java.awt.Color
import java.awt.Graphics2D
import java.awt.image.BufferedImage
import javax.imageio.ImageIO

data class Robot(var x: Int, var y: Int, val dx: Int, val dy: Int)

fun readInput(filename: String): List<String> {
    return File("$filename.txt").readLines()
}

fun buildRobots(config: List<String>): Triple<List<Robot>, Int, Int> {
    val robots = config.map { toRobot(it) }
    val tall = robots.maxOf { it.y } + 1
    val wide = robots.maxOf { it.x } + 1
    return Triple(robots, tall, wide)
}

fun toRobot(line: String): Robot {
    // p=2,4 v=2,-3
    val (p, v) = line.split(" ")
    val (x, y) = p.split("=")[1].split(",").map { it.toInt() }
    val (dx, dy) = v.split("=")[1].split(",").map { it.toInt() }
    return Robot(x, y, dx, dy)
}

fun moveRobots(robots: List<Robot>, moves: Int, wide: Int, tall: Int): List<Robot> {
    return robots.map { robot ->
        val newX = ((robot.x + robot.dx * moves) % wide + wide) % wide
        val newY = ((robot.y + robot.dy * moves) % tall + tall) % tall
        robot.copy(x = newX, y = newY)
    }
}

fun splitRegions(robots: List<Robot>, wide: Int, tall: Int): List<List<Robot>> {
    val midX = wide / 2
    val midY = tall / 2
    val regions = List(4) { mutableListOf<Robot>() }

    robots.forEach { robot ->
        val (x, y) = robot
        when {
            x == midX || y == midY -> return@forEach
            x < midX && y < midY -> regions[0].add(robot) // Top-left
            x >= midX && y < midY -> regions[1].add(robot) // Top-right
            x < midX && y >= midY -> regions[2].add(robot) // Bottom-left
            x >= midX && y >= midY -> regions[3].add(robot) // Bottom-right
        }
    }

    return regions
}

fun printRobots(robots: List<Robot>, wide: Int, tall: Int, print: Boolean = true): String {
    val grid = Array(tall) { CharArray(wide) { '.' } }
    robots.forEach { robot ->
        grid[robot.y][robot.x] = '#'
    }
    val pattern = grid.joinToString("\n") { it.joinToString("") }
    if (print) println("$pattern\n")
    return pattern
}

fun isUniquePositions(robots: List<Robot>): Boolean {
    val positions = robots.map { "${it.x},${it.y}" }.toSet()
    return positions.size == robots.size
}

fun writeSnapshot(robots: List<Robot>, wide: Int, tall: Int, seconds: Int, scale: Int = 10) {
    val grid = Array(tall) { CharArray(wide) { '.' } }
    robots.forEach { robot ->
        grid[robot.y][robot.x] = '#'
    }

    val baseImage  = BufferedImage(wide, tall, BufferedImage.TYPE_INT_RGB)

    for (y in 0 until tall) {
        for (x in 0 until wide) {
            val color = if (grid[y][x] == '.') Color.BLACK else Color.GREEN
            baseImage .setRGB(x, y, color.rgb)
        }
    }

    // Scale the image
    val scaledWidth = wide * scale
    val scaledHeight = tall * scale
    val scaledImage = BufferedImage(scaledWidth, scaledHeight, BufferedImage.TYPE_INT_RGB)
    val g2d: Graphics2D = scaledImage.createGraphics()

    // Draw the base image into the scaled image
    g2d.drawImage(baseImage, 0, 0, scaledWidth, scaledHeight, null)
    g2d.dispose()

    val filename = "images/frame_%04d.png".format(seconds)
    ImageIO.write(scaledImage, "png", File(filename))
}

fun part1(input: List<String>, moves: Int = 100, debug: Boolean = false): Int {
    val (robots, tall, wide) = buildRobots(input)
    val movedRobots = moveRobots(robots, moves, wide, tall)
    val regions = splitRegions(movedRobots, wide, tall)
    if (debug) {
        println(regions)
        printRobots(movedRobots, wide, tall)
    }
    return regions.fold(1) { factor, region -> factor * (region.size.takeIf { it > 0 } ?: 1) }
}

fun part2(input: List<String>, scale: Int = 10): Int {
    val (robots, tall, wide) = buildRobots(input)
    var currentRobots = robots
    val patternSet = mutableMapOf<String, Pair<Int, List<Robot>>>()
    var moves = 0

    while (true) {
        moves++
        currentRobots = moveRobots(currentRobots, 1, wide, tall)
        if (isUniquePositions(currentRobots)) {
            val pattern = printRobots(currentRobots, wide, tall, false)
            if (patternSet.containsKey(pattern)) {
                println("Found a loop! $moves moves")
                break
            }
            patternSet[pattern] = moves to currentRobots
            writeSnapshot(currentRobots, wide, tall, moves, scale)
        }
    }

    println("Total unique robot positions: ${patternSet.size}")
    patternSet.values.forEach { (moves, robots) ->
        println("Moves: $moves")
        printRobots(robots, wide, tall)
    }
    return 0
}

fun main() {
    val filename = "Day14"
    val input = readInput(filename)
    val inputTest = readInput("${filename}_test")

    println("Part 1: ${part1(input, debug = true)}") // 12
    println("Part 1: ${part1(inputTest, debug = true)}") // 218619120

    println("Part 2: ${part2(input)}") // 0
    println("Part 2: ${part2(inputTest)}") // 7055
}

File: 2024/14_Restroom_Redoubt/Day14.js
// Day 14: Restroom Redoubt
// https://adventofcode.com/2024/day/14

const fs = require('fs');

function readInput(filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8').split('\r\n');
}

function buildRobots(config) {
  const robots = config.map(toRobot);
  const tall = Math.max(...robots.map((r) => r.y)) + 1;
  const wide = Math.max(...robots.map((r) => r.x)) + 1;
  return [robots, tall, wide];
}

function toRobot(line) {
  // p=2,4 v=2,-3
  const [p, v] = line.split(' ');
  const [x, y] = p.split('=')[1].split(',').map(Number);
  const [dx, dy] = v.split('=')[1].split(',').map(Number);
  return {
    x,
    y,
    dx,
    dy,
  };
}

function moveRobots(robots, moves, wide, tall) {
  return robots.map((robot) => {
    const { x, y, dx, dy } = robot;
    let newX = (x + dx * moves) % wide;
    let newY = (y + dy * moves) % tall;

    newX = (newX + wide) % wide;
    newY = (newY + tall) % tall;

    return {
      x: newX,
      y: newY,
      dx,
      dy,
    };
  });
}

function splitRegions(robots, wide, tall) {
  const midX = wide >> 1;
  const midY = tall >> 1;
  const regions = [[], [], [], []];

  robots.forEach((robot) => {
    const { x, y } = robot;
    if (x === midX || y === midY) return;

    if (x < midX && y < midY) {
      regions[0].push(robot); // Top-left
    } else if (x >= midX && y < midY) {
      regions[1].push(robot); // Top-right
    } else if (x < midX && y >= midY) {
      regions[2].push(robot); // Bottom-left
    } else if (x >= midX && y >= midY) {
      regions[3].push(robot); // Bottom-right
    }
  });

  return regions;
}

function printRobots(robots, wide, tall, print = true) {
  const grid = Array.from({ length: tall }, () => Array(wide).fill('.'));
  robots.forEach((robot) => {
    const { x, y } = robot;
    grid[y][x] = '#';
  });
  const pattern = grid.map((row) => row.join('')).join('\n');
  if (print) console.log(pattern, '\n');
  return pattern;
}

function isUniquePositions(robots) {
  const positions = robots.map((robot) => `${robot.x},${robot.y}`);
  return new Set(positions).size === positions.length;
}

function part1({ input, moves = 100, debug = false }) {
  const [robots, tall, wide] = buildRobots(input);
  const movedRobots = moveRobots(robots, moves, wide, tall);
  const regions = splitRegions(movedRobots, wide, tall);
  if (debug) {
    console.log(regions);
    printRobots(movedRobots, wide, tall);
  }
  const safetyFactor = regions.reduce((factor, region) => factor * (region.length || 1), 1);
  return safetyFactor;
}

function part2({ input, debug = false }) {
  const buildResult = buildRobots(input);
  const [, tall, wide] = buildResult;
  let [robots] = buildResult;

  const patternSet = new Map();
  let moves = 0;
  while (true) {
    moves++;
    robots = moveRobots(robots, 1, wide, tall);

    if (isUniquePositions(robots)) {
      const pattern = printRobots(robots, wide, tall, false);
      if (patternSet.has(pattern)) {
        console.log('Found a loop!', moves, 'moves');
        break;
      }
      patternSet.set(pattern, [moves, robots]);
    }
  }

  console.log('Total unique robot positions:', patternSet.size);
  for (const [, [moves, robots]] of patternSet) {
    console.log('Moves:', moves);
    printRobots(robots, wide, tall);
  }
  return 0;
}

const filename = 'Day14';
const input = readInput(filename);
const inputTest = readInput(`${filename}_test`);

console.log('Part 1:', part1({ input, debug: true })); // 12
console.time('Part 1');
console.log('Part 1 Test:', part1({ input: inputTest })); // 218619120
console.timeEnd('Part 1');

console.log('Part 2:', part2({ input, debug: true })); // 0
console.time('Part 2');
console.log('Part 2 Test:', part2({ input: inputTest })); // 7055
console.timeEnd('Part 2');
